### PR TITLE
gRPC Client

### PR DESCRIPTION
**Pull request by aelred**
# Why is this PR needed?

We need a Java gRPC client to use our gRPC server - this is our thin/dumb client.

# What does the PR do?

Creates alternative implementations of interfaces such as `GraknTx` and `QueryRunner` that execute using an existing Grakn server.

# Does it break backwards compatibility?

No.

# List of future improvements not on this PR

Loads! There are many methods that simply `throw UnsupportedOperationException()`.

We also will need to make the `Grakn#session` method start a `RemoteGraknSession`, instead of a `(Local)GraknSessionImpl`. For now, I have a different entry point: `RemoteGrakn#session`.

### CODE REVIEW

File: grakn-client/src/main/java/ai/grakn/remote/RemoteGrakn.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+package ai.grakn.remote;
+
+import ai.grakn.GraknSession;
+import ai.grakn.Keyspace;
+import ai.grakn.util.SimpleURI;
+
+/**
+ * @author Felix Chapman
+ */
+public final class RemoteGrakn {
```
Comment (fppt) [inline grakn-client/src/main/java/ai/grakn/remote/RemoteGrakn.java]: Java docs need work here.

File: grakn-client/src/main/java/ai/grakn/remote/GrpcClient.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    private TxResponse responseOrThrow() {
+        Response response;
+
+        try {
+            response = communicator.receive();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw Throwables.propagate(e);
```
Comment (fppt) [inline grakn-client/src/main/java/ai/grakn/remote/GrpcClient.java]: What does this do? Also: https://github.com/google/guava/wiki/Why-we-deprecated-Throwables.propagate . . . Relevant?

File: grakn-client/src/main/java/ai/grakn/remote/RemoteGraknSession.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+import ai.grakn.rpc.generated.GraknGrpc.GraknStub;
+import ai.grakn.util.SimpleURI;
+import com.google.common.annotations.VisibleForTesting;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+
+/**
+ * @author Felix Chapman
+ */
+class RemoteGraknSession implements GraknSession {
```
Comment (fppt) [inline grakn-client/src/main/java/ai/grakn/remote/RemoteGraknSession.java]: Ditto for docs

File: grakn-client/src/main/java/ai/grakn/remote/RemoteGraknSession.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    @Override
+    public GraknTx open(GraknTxType transactionType) {
+        return RemoteGraknTx.create(this, transactionType);
+    }
+
+    @Override
+    public GraknComputer getGraphComputer() {
+        throw new UnsupportedOperationException(); // TODO
```
Comment (fppt) [inline grakn-client/src/main/java/ai/grakn/remote/RemoteGraknSession.java]: Ohhhhh this is not going to be fun

File: grakn-client/src/main/java/ai/grakn/remote/RemoteGraknTx.java [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Stream;
+
+/**
+ * @author Felix Chapman
+ */
+class RemoteGraknTx implements GraknTx, GraknAdmin {
```
Comment (fppt) [inline grakn-client/src/main/java/ai/grakn/remote/RemoteGraknTx.java]: Ditto for docs

File: grakn-client/src/main/java/ai/grakn/remote/RemoteGraknSession.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    @Override
+    public Keyspace keyspace() {
+        return keyspace;
+    }
+
+    @Override
+    public GraknConfig config() {
+        throw new UnsupportedOperationException(); // TODO
```
Comment (fppt) [inline grakn-client/src/main/java/ai/grakn/remote/RemoteGraknSession.java]: This should be "easy" enough to do. It just needs to mimic the behaviour of the REST controller used to get the config. I.e. read the properties file locally and overwrite some keys.

File: grakn-client/src/main/java/ai/grakn/remote/RemoteQueryRunner.java [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * @author Felix Chapman
+ */
+final class RemoteQueryRunner implements QueryRunner {
```
Comment (fppt) [inline grakn-client/src/main/java/ai/grakn/remote/RemoteQueryRunner.java]: 🐐

File: grakn-client/src/main/java/ai/grakn/remote/GrpcClient.java (hunk 2 above)
Comment (aelred) [reply grakn-client/src/main/java/ai/grakn/remote/GrpcClient.java]: Ooh... I didn't know that. So it looks like I should just say `throw new RuntimeException(e);`. I hate `InterruptedException`s 😞

File: grakn-client/src/main/java/ai/grakn/remote/RemoteGraknSession.java (hunk 6 above)
Comment (aelred) [reply grakn-client/src/main/java/ai/grakn/remote/RemoteGraknSession.java]: If this is a remote API, then we can't expect the config file to be present. I think we should try removing this method from `GraknSession` and make it internal - but that's easier said than done 😛

File: grakn-engine/src/test/java/ai/grakn/engine/rpc/GrpcServerTest.java [hunk 8]
```diff
@@ -83,9 +88,7 @@
 public class GrpcServerTest {
 
     private static final String EXCEPTION_MESSAGE = "OH DEAR";
-    private static final GraknException EXCEPTION = new GraknException(EXCEPTION_MESSAGE) {
-        private static final long serialVersionUID = 8491972914457727509L;
-    };
+    private static final GraknException EXCEPTION = GraqlQueryException.create(EXCEPTION_MESSAGE);
```
Comment (duckofyork) [inline grakn-engine/src/test/java/ai/grakn/engine/rpc/GrpcServerTest.java]: I like the exception message 👍

File: grakn-client/src/main/java/ai/grakn/remote/RemoteGraknSession.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        this.channel = channel;
+    }
+
+    @VisibleForTesting
+    public static RemoteGraknSession create(Keyspace keyspace, SimpleURI uri, ManagedChannel channel) {
+        return new RemoteGraknSession(keyspace, uri, channel);
+    }
+
+    public static RemoteGraknSession create(Keyspace keyspace, SimpleURI uri){
+        // TODO: usePlainText is insecure
```
[the remark this answers is not in the document]
Comment (aelred) [reply grakn-client/src/main/java/ai/grakn/remote/RemoteGraknSession.java]: It means we're explicitly not using encryption - like http vs https. I'll add a comment explaining.

**Pull request merged**